arenadata_toolkit: add skew coefficients for non-leaf partitions

red1452 opened the pull request:
Before this patch `adb_skew_coefficients` worked only with the last level of partitions correctly. For other parent partitions and root of partitions `adb_skew_coefficients` can not show coefficients because select sizes from `adb_relation_storage_size` returns sizes only for last level of partitions.

View `adb_skew_coefficients` was rewrote. New version `adb_skew_coefficients` calculates sizes for all levels of partitions and generates table with these sizes for each segments. It uses recursive CTE to get sizes from lowest level of partitioned table to the root. Table with sizes is used for calculate skew coefficients as it is done at previous version.

Warning: before this patch filter by "oid" was applied at first step (before executing `adb_relation_storage_size`) of query plan. Because of this, the query with the filter was executed much faster than the full selection (without filter). At new version of `adb_skew_coefficients`  filter by "oid" is applied at the last step of query plan. Time, what is need for executing query with filter, is the same as for query without filter.

Name of table `db_files_history_1_prt_p202306` in test `arenadata_toolkit_test` was changed to `db_files_history_1_prt_pYYYYMM`, because this date depends on current month and year.

## Performance comparison

`select * from arenadata_toolkit.adb_skew_coefficients;`
```
------------------------------------------------------------------------
| attempt  | adb_skew_coefficients before  | adb_skew_coefficients new |
------------------------------------------------------------------------
| 1        | 77551 ms                      | 73822 ms                  |
------------------------------------------------------------------------
| 2        | 69888 ms                      | 70008 ms                  |
------------------------------------------------------------------------
| 3        | 66094 ms                      | 70298 ms                  |
------------------------------------------------------------------------
| 4        | 66379 ms                      | 64485 ms                  |
------------------------------------------------------------------------
```

red1452 committed:
arenadata_toolkit: add support for working with partitions to adb_skew_coefficients (#607)

Before this patch adb_skew_coefficients worked only with the last level
of partitions correctly. For other parent partitions and root of partitions
adb_skew_coefficients can not show coeffcients because select sizes
from adb_relation_storage_size returns sizes only for last level of partitions.

View adb_skew_coefficients was rewrited to work with new function get_full_storage_size.
Function get_full_storage_size calculates sizes for all levels of partitions and returns
table with sizes.

Version of arenadata_toolkit was increased to 1.1.

InnerLife0 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0-1.1.sql:
```diff
@@ -0,0 +1,83 @@
+						FROM gp_dist_random('gp_toolkit.__gp_user_tables')
+						WHERE autrelstorage != 'x')
+				SELECT *,
+						ROW_NUMBER () OVER (
```
Are you sure that gp_dist_random() always gives you the strict order, giving you the full set of data from segment and step by step from 1 to n segment?

red1452 replied:
Strict order is not important here, because we use `stddev(size)` and `avg(size)`, when we want to get result of `adb_skew_coefficients`. `ROW_NUMBER` needs to get only difference between segments at relations with the same name.

InnerLife0 replied:
Are you sure that you can't calc sum() for partitions from different segments?

red1452 replied:
Before calculate `stddev(size)` and `avg(size)`, I must get sizes of tables for each segment. Intermediate table looks like:
```
  id   |                          relname                          |  size   | seg_id 
-------+-----------------------------------------------------------+---------+--------
 27472 | part_table                                                | 1572864 |      2
 27472 | part_table                                                |  983040 |      3
 27472 | part_table                                                |  983040 |      1
 27478 | part_table_1_prt_1                                        |  524288 |      1
 27478 | part_table_1_prt_1                                        | 1114112 |      2
 27478 | part_table_1_prt_1                                        |  524288 |      3
 27486 | part_table_1_prt_1_2_prt_1                                |  262144 |      1
 27486 | part_table_1_prt_1_2_prt_1                                |  851968 |      2
 27486 | part_table_1_prt_1_2_prt_1                                |  262144 |      3
 27497 | part_table_1_prt_1_2_prt_1_3_prt_1                        |  131072 |      3
 27497 | part_table_1_prt_1_2_prt_1_3_prt_1                        |  720896 |      2
 27497 | part_table_1_prt_1_2_prt_1_3_prt_1                        |  131072 |      1
 27509 | part_table_1_prt_1_2_prt_1_3_prt_1_4_prt_1                |  655360 |      2
 27509 | part_table_1_prt_1_2_prt_1_3_prt_1_4_prt_1                |   65536 |      3
 27509 | part_table_1_prt_1_2_prt_1_3_prt_1_4_prt_1                |   65536 |      1
 27522 | part_table_1_prt_1_2_prt_1_3_prt_1_4_prt_1_5_prt_sub_prt1 |   32768 |      1
 27522 | part_table_1_prt_1_2_prt_1_3_prt_1_4_prt_1_5_prt_sub_prt1 |   32768 |      3
 27522 | part_table_1_prt_1_2_prt_1_3_prt_1_4_prt_1_5_prt_sub_prt1 |  622592 |      2
 27536 | part_table_1_prt_1_2_prt_1_3_prt_1_4_prt_1_5_prt_sub_prt2 |   32768 |      3
 27536 | part_table_1_prt_1_2_prt_1_3_prt_1_4_prt_1_5_prt_sub_prt2 |   32768 |      2
 27536 | part_table_1_prt_1_2_prt_1_3_prt_1_4_prt_1_5_prt_sub_prt2 |   32768 |      1
...
```
Field `seg_id` helps to separate values from different segments.

InnerLife0 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0-1.1.sql:
```diff
@@ -0,0 +1,83 @@
+		$fmt$);
+	END LOOP;
+
+	RETURN QUERY SELECT * FROM __adb_full_storage;
```
Is it possible to return this without intermediate table? Using window functions, recursive queries or etc?

red1452 replied:
I think it is not possible, because we need to update a part of rows before returning.

InnerLife0 replied:
This is a rollup total from leafs to root. If we need to rollup only deepest leafs, then query will be simpler.
```
with recursive cte as (
	select relname nm, relname::regclass::oid id, inhparent parent_id, length(relname) self_size
	from pg_class t1
	left join pg_inherits t2 on inhrelid = relname::regclass::oid
	where relname like 'part_table%'
	UNION ALL
	select relname nm, relname::regclass::oid id, inhparent parent_id, self_size
	from pg_class t1
	left join pg_inherits t2 on inhrelid = relname::regclass::oid
	join cte on cte.parent_id = relname::regclass::oid
	where relname like 'part_table%'
)
select nm, sum(self_size) total, length(nm) self_size
from
cte
group by nm
order by 1
```

red1452 replied:
I rewrote view by using your query, thank you!

InnerLife0 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0-1.1.sql:
```diff
@@ -0,0 +1,83 @@
+					FROM tmp_size;
+	$fmt$);
+
+	FOR r IN 1..3 LOOP
```
What is the magic '3' number for? Why do we need to do same loop triple times?

red1452 replied:
3 is maximum of level of partitions. One `UPDATE` updates size of one level of partitions. Need to execute three times this `UPDATE` to update all levels of partitions.

InnerLife0 replied:
Check this one:
```
CREATE TABLE part_table (id int, a int, b int, c int, d int, str text)
	DISTRIBUTED BY (id)
	PARTITION BY RANGE (a)
		SUBPARTITION BY RANGE (b)
			SUBPARTITION TEMPLATE (
					START (1) END (3) EVERY (1))
				SUBPARTITION BY RANGE (c) SUBPARTITION TEMPLATE (START (1) END (3) EVERY (1))
				SUBPARTITION BY RANGE (d) SUBPARTITION TEMPLATE (START (1) END (3) EVERY (1))
				SUBPARTITION BY LIST (str)
					SUBPARTITION TEMPLATE (
						SUBPARTITION sub_prt1 VALUES ('sub_prt1'),
						SUBPARTITION sub_prt2 VALUES ('sub_prt2'))
(START (1) END (3) EVERY (1));
```

red1452 replied:
I added this table to test.

InnerLife0 commented on gpcontrib/arenadata_toolkit/expected/arenadata_toolkit_skew_test.out:
```diff
@@ -1,4 +1,5 @@
 CREATE EXTENSION arenadata_toolkit;
+SET client_min_messages=WARNING;
```
What messages are now hided by this SET?

red1452 replied:
This SET hides messages, which generated at `DROP IF NOT EXIST` and `CREATE TEMPORARY TABLE` at new function `get_full_storage_size`.

InnerLife0 replied:
Then we need to left a comment here, about why we suppress log messages.
 IMO, if we really need it, it's better to place it near the blocks which can generate such message and reset it just after them.

red1452 replied:
Removed, because I removed function `get_full_storage_size`.

red1452 committed:
add autogenerated sql script to gitignore

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53275

red1452 committed:
Revert "add autogenerated sql script to gitignore"

This reverts commit 1287858ba1dd7380fbe919edf8468b9419ebea14.

red1452 committed:
Revert "arenadata_toolkit: add support for working with partitions to adb_skew_coefficients (#607)"

This reverts commit a2d4df32208dd223ef4cba45d47dbf4dc7493d65.

red1452 committed:
Merge remote-tracking branch 'origin/adb-6.x-dev' into ADBDEV-4070

red1452 committed:
arenadata_toolkit: add support for working with partitions to adb_skew_coefficients (#607)

Before this patch adb_skew_coefficients worked only with the last level of partitions correctly.
For other parent partitions and root of partitions adb_skew_coefficients can not show coeffcients
because select sizes from adb_relation_storage_size returns sizes only for last level of partitions.

View adb_skew_coefficients was rewrited to work with new function get_full_storage_size.
Function get_full_storage_size calculates sizes for all levels of partitions and returns table with sizes.

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4070

red1452 committed:
remove extra changing client_min_messages

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53845

dnskvlnk commented:
Seems like you deleted `get_full_storage_size`, but there are references of it that remained in PR and commit description

red1452 commented:
> Seems like you deleted `get_full_storage_size`, but there are references of it that remained in PR and commit description

New commit message:
```
arenadata_toolkit: add support for working with partitions to adb_skew_coefficients (#607)

Before this patch adb_skew_coefficients worked only with the last level of partitions correctly.
For other parent partitions and root of partitions adb_skew_coefficients can not show coefficients
because select sizes from adb_relation_storage_size returns sizes only for last level of partitions.

View adb_skew_coefficients was rewrote. New version adb_skew_coefficients calculates sizes for all levels
of partitions and returns table with sizes. It uses recursive CTE to get sizes from lowest level of partitioned
table to the root. Table with sizes is used for calculate skew coefficients as it is done at previous version.
```

dnskvlnk commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,41 @@
+				arenadata_toolkit.adb_relation_storage_size(t1.autoid) AS size
+			FROM gp_dist_random('gp_toolkit.__gp_user_tables') t1
+			LEFT JOIN pg_inherits t2 ON inhrelid = t1.autoid
+			WHERE autrelstorage != 'x')
```
Do we need 'v' and 'f' relation types?

red1452 replied:
I think, we need all types of relation, which can use filesystem.

RekGRpth commented:
> Performance comparison

May be use explain analyze to show memory consumption?

red1452 committed:
optimization of adb_skew_coefficients

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4070

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54364

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,40 @@
+		SELECT cte.parent_id id, cte.seg_id, cte.size, inh.inhparent parent_id
+			FROM cte
+			LEFT JOIN pg_inherits inh ON inhrelid = cte.parent_id
+			WHERE cte.parent_id != 0
```
```suggestion
			WHERE inhrelid IS NOT NULL
```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,40 @@
+			pgc.oid id,
+			pgc.gp_segment_id seg_id,
+			arenadata_toolkit.adb_relation_storage_size(pgc.oid) size,
+			inh.inhparent parent_id
```
```suggestion
			inhparent parent_id
```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,40 @@
+			arenadata_toolkit.adb_relation_storage_size(pgc.oid) size,
+			inh.inhparent parent_id
+		FROM gp_dist_random('pg_class') pgc
+		LEFT JOIN pg_inherits inh ON inh.inhrelid = pgc.oid
```
```suggestion
		LEFT JOIN pg_inherits ON inhrelid = pgc.oid
```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,40 @@
+		WHERE pgc.relkind = 'r' AND pgc.relstorage != 'x' AND
+			pgc.relnamespace IN (SELECT aunoid FROM gp_toolkit.__gp_user_namespaces)
+	UNION ALL
+		SELECT cte.parent_id id, cte.seg_id, cte.size, inh.inhparent parent_id
```
```suggestion
		SELECT inhrelid id, seg_id, size, inhparent parent_id
```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,40 @@
+	UNION ALL
+		SELECT cte.parent_id id, cte.seg_id, cte.size, inh.inhparent parent_id
+			FROM cte
+			LEFT JOIN pg_inherits inh ON inhrelid = cte.parent_id
```
```suggestion
			LEFT JOIN pg_inherits ON inhrelid = parent_id
```

RekGRpth commented:
Additional simplify
```sql
...
	SELECT
			pgc.oid id,
			pgc.gp_segment_id seg_id,
			arenadata_toolkit.adb_relation_storage_size(pgc.oid) size,
			NULL::oid parent_id
		FROM gp_dist_random('pg_class') pgc
		WHERE pgc.relkind = 'r' AND pgc.relstorage != 'x' AND
			pgc.relnamespace IN (SELECT aunoid FROM gp_toolkit.__gp_user_namespaces)
	UNION ALL
		SELECT inhparent id, seg_id, size, inhrelid parent_id
			FROM cte
			LEFT JOIN pg_inherits ON inhrelid = id
			WHERE inhparent IS NOT NULL
...
```

RekGRpth commented:
also on standup it was suggested using wrapper
```sql
CREATE FUNCTION arenadata_toolkit.adb_relation_storage_size_segment_location(reloid OID, forkName TEXT default 'main')
RETURNS BIGINT
AS 'SELECT arenadata_toolkit.adb_relation_storage_size($1, $2)'
LANGUAGE SQL VOLATILE STRICT EXECUTE ON ALL SEGMENTS;

```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,40 @@
+			WHERE cte.parent_id != 0
+	),
+	tables_size_by_segments AS (
+		SELECT id, sum(size) as size, seg_id
```
```suggestion
		SELECT id, sum(size) as size
```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,40 @@
+	tables_size_by_segments AS (
+		SELECT id, sum(size) as size, seg_id
+			FROM cte
+			GROUP BY (id, seg_id)),
```
```suggestion
			GROUP BY id, seg_id),
```

RekGRpth commented:
> ```sql
> EXECUTE ON ALL SEGMENTS
> ```

or more simply
```sql
CREATE FUNCTION arenadata_toolkit.adb_relation_storage_size_segment_location(reloid OID, forkName TEXT default 'main')
RETURNS BIGINT
AS '$libdir/arenadata_toolkit', 'adb_relation_storage_size'
LANGUAGE C VOLATILE STRICT EXECUTE ON ALL SEGMENTS;
```

red1452 committed:
optimization of query, add 'execute on all segments to adb_relation_storage_size'

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4070

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,45 @@
 
 ALTER FUNCTION arenadata_toolkit.adb_get_relfilenodes(tablespace_oid OID) ROWS 30000000;
+
+CREATE OR REPLACE FUNCTION arenadata_toolkit.adb_rel_storage_size_on_seg(reloid OID, forkName TEXT default 'main')
```
I think you should call this function adb_relation_storage_size_segment_location.

red1452 replied:
I minimized length of function's name to minimize difference in the tests output.

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,45 @@
+
+REVOKE EXECUTE ON FUNCTION arenadata_toolkit.adb_rel_storage_size_on_seg(OID, TEXT) FROM public;
+
+CREATE OR REPLACE VIEW arenadata_toolkit.adb_skew_coefficients
```
Could you reformat this view like the first version of this view?

RekGRpth commented on gpcontrib/arenadata_toolkit/expected/arenadata_toolkit_test.out:
```diff
@@ -93,6 +93,7 @@ SELECT objname, objtype, objstorage, objacl FROM toolkit_objects_info ORDER BY o
  adb_collect_table_stats             | proc    | -          | {owner=X/owner}
  adb_create_tables                   | proc    | -          | {owner=X/owner}
  adb_get_relfilenodes                | proc    | -          | {=X/owner,owner=X/owner}
+ adb_rel_storage_size_on_seg         | proc    | -          | {owner=X/owner}
```
Why are the values in the last column in this row and the next one different?

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54406

red1452 committed:
add adb_relation_storage_size_on_segments

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4070

red1452 commented:
New commit message:
```
arenadata_toolkit: add skew coefficients for non-leaf partitions (#607)

Before this patch 'adb_skew_coefficients' worked only with the last level of
partitions correctly. For other parent partitions and root of partitions
'adb_skew_coefficients' can not show coefficients because select sizes from
'adb_relation_storage_size' returns sizes only for last level of partitions.

View 'adb_skew_coefficients' was rewrote. New version 'adb_skew_coefficients'
calculates sizes for all levels of partitions and generates table with these
sizes for each segments. It uses recursive CTE to get sizes from lowest level
of partitioned table to the root. Table with sizes is used for calculate skew
coefficients as it is done at previous version.

Warning: before this patch filter by "oid" was applied at first step (before
executing 'adb_relation_storage_size') of query plan. Because of this, the
query with the filter was executed much faster than the full selection (without
filter). At new version of 'adb_skew_coefficients'  filter by "oid" is applied
at the last step of query plan. Time, what is need for executing query with
filter, is the same as for query without filter.
```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,50 @@
+	WHERE relkind = 'r' AND relstorage != 'x' AND
+		relnamespace IN (SELECT aunoid FROM gp_toolkit.__gp_user_namespaces)
+	UNION ALL
+		SELECT inhparent AS id, seg_id, size
```
I think lines 26-29 should be moved further to the left.

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,50 @@
+	UNION ALL
+		SELECT inhparent AS id, seg_id, size
+		FROM cte
+		LEFT JOIN pg_inherits ON inhrelid = id
```
```suggestion
		LEFT JOIN pg_catalog.pg_inherits ON inhrelid = id
```

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54679

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,50 @@
+		stddev(size) AS skewdev,
+		avg(size) AS skewmean
+	FROM tables_size_by_segments
+	GROUP BY id)
```
```suggestion
	GROUP BY id
)
```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.0--1.1.sql:
```diff
@@ -1,3 +1,50 @@
+	skew.skewoid AS skcoid,
+	pgn.nspname  AS skcnamespace,
+	pgc.relname  AS skcrelname,
+	CASE WHEN skewdev > 0 THEN skewdev/skewmean * 100.0 ELSE 0 END AS skccoeff
```
Why is the numerator checked for zero, and not the denominator? It is clear that if the numerator is not zero, then neither is the denominator, but still...

red1452 replied:
Because if table has same sizes at all segments (for example 32768) than skewmean is 32768 and skewdev is 0. You can do `skewdev/skewmean * 100.0` but you will get 0. Why do we have to do extra calculations (It will be if change `skewdev > 0` to `skewmean > 0`)?

RekGRpth commented:
> select sizes from
> 'adb_relation_storage_size' returns sizes only for last level of partitions

Why did you decide not to have the adb_relation_storage_size function return the total size of all partition levels?

red1452 commented:
> > select sizes from
> > 'adb_relation_storage_size' returns sizes only for last level of partitions
> 
> Why did you decide not to have the adb_relation_storage_size function return the total size of all partition levels?

`adb_relation_storage_size` returns size of files which belong to relation. Non-leaf partitions do not have files. `adb_relation_storage_size` does not make: which partitions this relation has. At [specification](https://wiki.adsw.io/books/adb/page/arenadata-toolkit-adb-skew-coefficients-partitioned-tables-improvement) do not say anything about changing `adb_relation_storage_size`, only `adb_skew_coefficients`.

red1452 committed:
cosmetic fixes

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4070

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54739

red1452 committed:
arenadata_toolkit: add skew coefficients for non-leaf partitions (#607)

Before this patch 'adb_skew_coefficients' worked only with the last level of
partitions correctly. For other parent partitions and root of partitions
'adb_skew_coefficients' can not show coefficients because select sizes from
'adb_relation_storage_size' returns sizes only for last level of partitions.

View 'adb_skew_coefficients' was rewrote. New version 'adb_skew_coefficients'
calculates sizes for all levels of partitions and generates table with these
sizes for each segments. It uses recursive CTE to get sizes from lowest level
of partitioned table to the root. Table with sizes is used for calculate skew
coefficients as it is done at previous version.

Warning: before this patch filter by "oid" was applied at first step (before
executing 'adb_relation_storage_size') of query plan. Because of this, the
query with the filter was executed much faster than the full selection (without
filter). At new version of 'adb_skew_coefficients'  filter by "oid" is applied
at the last step of query plan. Time, what is need for executing query with
filter, is the same as for query without filter.

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4070

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/55503

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/665899

red1452 committed:
small cosmetic fix in test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/55526

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/666647

red1452 committed:
another small fix in test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/55550

red1452 committed:
arenadata_toolkit: add skew coefficients for non-leaf partitions (#607)

Before this patch 'adb_skew_coefficients' worked only with the last level of
partitions correctly. For other parent partitions and root of partitions
'adb_skew_coefficients' can not show coefficients because select sizes from
'adb_relation_storage_size' returns sizes only for last level of partitions.

View 'adb_skew_coefficients' was rewrote. New version 'adb_skew_coefficients'
calculates sizes for all levels of partitions and generates table with these
sizes for each segments. It uses recursive CTE to get sizes from lowest level
of partitioned table to the root. Table with sizes is used for calculate skew
coefficients as it is done at previous version.

Warning: before this patch filter by "oid" was applied at first step (before
executing 'adb_relation_storage_size') of query plan. Because of this, the
query with the filter was executed much faster than the full selection (without
filter). At new version of 'adb_skew_coefficients'  filter by "oid" is applied
at the last step of query plan. Time, what is need for executing query with
filter, is the same as for query without filter.

Name of table 'db_files_history_1_prt_p202306' in test 'arenadata_toolkit_test'
was changed to 'db_files_history_1_prt_pYYYYMM', because this date depends on
current month and year.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/55573

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/668564
